synapse/util/caches/lrucache.py

```diff
@@ -57,13 +59,54 @@ class _Node:
     __slots__ = ["prev_node", "next_node", "key", "value", "callbacks"]
 
     def __init__(
-        self, prev_node, next_node, key, value, callbacks: Optional[set] = None
+        self,
+        prev_node,
+        next_node,
+        key,
+        value,
+        callbacks: Collection[Callable[[], None]] = (),
     ):
         self.prev_node = prev_node
         self.next_node = next_node
         self.key = key
         self.value = value
-        self.callbacks = callbacks or set()
+
+        # Set of callbacks to run when the node gets deleted. We store as a list
+        # rather than a set to keep memory usage down (and since we expect few
+        # entries per node the performance of checking for duplication in a list
+        # vs using a set is negligible).
+        #
+        # Note that we store this as an optional list to keep the memory
+        # footprint down. Empty lists are 56 bytes (and empty sets are 216 bytes).
+        self.callbacks = None  # type: Optional[List[Callable[[], None]]]
```

Comment:
We should probably be talking about the memory usage of empty lists versus `None` here, instead of vs. Sets?

Reply:
I'll clarify, but `None` as a singleton is basically free. The comparison with an empty set is because using sets is the "obvious" way of doing it.

Reply:
Right, OK. I read it as a justification for why you're using an optional list vs. just a list, rather than an optional list vs. a set :)